src/main/scala/is/hail/methods/LinearRegression3.scala

```diff
@@ -0,0 +1,134 @@
+package is.hail.methods
+
+import breeze.linalg._
+import breeze.numerics.sqrt
+import is.hail.annotations.Annotation
+import is.hail.expr._
+import is.hail.io.bgen.Bgen12GenotypeIterator
+import is.hail.stats._
+import is.hail.utils._
+import is.hail.variant._
+import net.sourceforge.jdistlib.T
+
+object LinearRegression3 {
+  def schema = TStruct(
+    ("beta", TArray(TDouble)),
+    ("se", TArray(TDouble)),
+    ("tstat", TArray(TDouble)),
+    ("pval", TArray(TDouble)))
+
+  def apply(vds: VariantDataset, ysExpr: Array[String], covExpr: Array[String], root: String, useDosages: Boolean, variantBlockSize: Int): VariantDataset = {
+    require(vds.wasSplit)
+
+    val (y, cov, completeSamples) = RegressionUtils.getPhenosCovCompleteSamples(vds, ysExpr, covExpr)
+    val completeSamplesSet = completeSamples.toSet
+    val sampleMask = vds.sampleIds.map(completeSamplesSet).toArray
+    val completeSampleIndex = (0 until vds.nSamples)
+      .filter(i => completeSamplesSet(vds.sampleIds(i)))
+      .toArray
+
+    val n = y.rows
+    val k = cov.cols
+    val d = n - k - 1
+    val dRec = 1d / d
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    info(s"Running linear regression for ${ y.cols } ${ plural(y.cols, "phenotype") } on $n samples with $k ${ plural(k, "covariate") } including intercept...")
+
+    val Qt = qr.reduced.justQ(cov).t
+    val Qty = Qt * y
+
+    val sc = vds.sparkContext
+    val sampleMaskBc = sc.broadcast(sampleMask)
+    val completeSampleIndexBc = sc.broadcast(completeSampleIndex)
+    val yBc = sc.broadcast(y)
+    val QtBc = sc.broadcast(Qt)
+    val QtyBc = sc.broadcast(Qty)
+    val yypBc = sc.broadcast(y.t(*, ::).map(r => r dot r) - Qty.t(*, ::).map(r => r dot r))
+
+    val pathVA = Parser.parseAnnotationRoot(root, Annotation.VARIANT_HEAD)
+    val (newVAS, inserter) = vds.insertVA(LinearRegressionMultiPheno.schema, pathVA)
+
+    val newRDD = vds.rdd.mapPartitions({ it =>
+      val missingSamples = new ArrayBuilder[Int]
+
+      // columns are genotype vectors
+      var X: DenseMatrix[Double] = null
+
+      it.grouped(variantBlockSize)
+        .flatMap { git =>
+          val a = git.toArray
+          val l = a.length
+
+          if (X == null || X.cols != l)
+            X = new DenseMatrix[Double](n, l)
+
+          var i = 0
+          while (i < l) {
+            val (_, (_, gs)) = a(i)
+
+            gs match {
```

Comment:
if we want to represent things that way on disk / in memory, we need to make sure that the nice interface is preserved. We still have to be able to look at the gt and express the dosage easily